### DISCUSSION TITLE
React vs Vue vs Solid

### DISCUSSION THREAD

**Discussion by brillout** [vikejs/bati#85] [Misc]
In the following, Bati's offical recommendation.

The recommendation follows an open discussion: everyone (especially UI framework maintainers) can engage in a conversation with us if they believe our recommendation to not be accurate.

> All three are great choices and there are plenty of companies proving that you can build a successful business with each.
>
> A lot of the choice is about personal preference: we recommend reading the offical website and documentation of [React](https://react.dev/), [Vue](https://vuejs.org/), and [Solid](https://www.solidjs.com/) and see what resonates most with you. (If all three seem just fine to you and none of them sticks out, then you can fallback with the conservative choice of picking the most widespread one which currently is React.)
>
> Vike works with any UI framework, which means that you can use several UI frameworks for the same website. For example, you can use Vue or Solid while using this one React library that only works with React. (That's why we say that "Vike has the largest ecossystem" as you can cherry-pick from the ecosystem of any UI framework.)
>
> Instead of getting stuck with decision paralysis, we recommend to decide and get started quickly and, as you scale, see for yourself whether you like your initial choice, and [progressively switch UI frameworks]() if you change your mind.
>
> React:
> - Pro: Largest ecosystem.
> - Pro: JSX templates (personal preference: it's a "con" for some people).
> - Pro: Innovative.
>   - React introduced many novelties, most notably unidirectional UIs, JSX, Hooks, and Server Components.
> - Con: Sometimes difficult to use.
>   - Its reactivity system is based on immutable objects which makes React inherently difficult to use (for example, you'll need to use [`memo`](https://react.dev/reference/react/memo) to keep your app performant).
>   - Sometimes complex, for example [UI tearing](https://github.com/reactwg/react-18/discussions/69) is complex in a leaky way: libraries cannot always fully abtract that complexity away.
>   - It has to be shown whether [RSC](https://react.dev/blog/2023/03/22/react-labs-what-we-have-been-working-on-march-2023#react-server-components) will manage to provide an easy DX. Today, using RSC is difficult to use.
>   - It has a lot of paper cuts, such as unhelpful error messages, no built-in TypeScript types, or libraries with broken ESM which need manual fixing.
>
> Vue:
> - Pro: Large ecosystem.
> - Pro: Simple.
> - Pro: HTML-like templates (personal preference: it's a "con" for some people).
> - Pro: Cohesive ecosystem of basic tools.
>
> Solid:
> - Pro: Simple.
>   - Its reactivity system is based on observable objects, which makes it considerably simpler to use and more performant.
> - Pro: Highly performant.
> - Pro: JSX templates (personal preference: it's a "con" for some people).
> - Pro: Responsive maintainers and bugs are fixed promptly.
> - Con: Its ecosystem is growing but still not as large as React/Vue's ecosystem.

**Comment by magne4000** [comment]
Really good points. The next task on my todo list was to explain the recommendations, so thanks for that 😃
I still need to find a proper way to show them in the interface (could be difficult for the embed/widget UI)

**Comment by brillout** [comment]
You're welcome 😀 Although it's a rough draft that needs polishing.

**Comment by brillout** [comment]
> needs polishing

Speaking of which:

> ~Beyond personal preference, there is one aspect we recommend to be careful about: the ecosystem. We recommend to think about what tools your business will need and to double check whether the UI framework's ecosystem has what you need.~
>
> Vike works with any UI framework, which means that you can use several UI frameworks for the same website. For example, you can use Vue or Solid while using this one React library that only works with React. (That's why we say that "Vike has the largest ecossystem" as you can cherry-pick from the ecosystem of any UI framework.)
>
> Instead of getting stuck with decision paralysis, we recommend to decide and get started quickly and, as you scale, see for yourself whether you like your initial choice, and [progressively switch UI frameworks]() if you change your mind.

**Comment by brillout** [comment]
(I'll finish my rewrite after I've released the V1 design and deprecated the current design.)
